File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/statistikk/sykefraværshistorikk/aggregert/Trendkalkulator.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.aggregert;

import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.UmaskertSykefraværForEttKvartal.hentUtKvartal;

import io.vavr.control.Either;
import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;
import lombok.AllArgsConstructor;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.ÅrstallOgKvartal;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.StatistikkException;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.UmaskertSykefraværForEttKvartal;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.UtilstrekkeligDataException;

@AllArgsConstructor
class Trendkalkulator {

  List<UmaskertSykefraværForEttKvartal> datagrunnlag;
  ÅrstallOgKvartal sistePubliserteKvartal;


  Either<UtilstrekkeligDataException, Trend> kalkulerTrend() {

    ÅrstallOgKvartal ettÅrSiden = sistePubliserteKvartal.minusEttÅr();

    Optional<UmaskertSykefraværForEttKvartal> nyesteSykefravær =
        hentUtKvartal(datagrunnlag, sistePubliserteKvartal);
    Optional<UmaskertSykefraværForEttKvartal> sykefraværetEttÅrSiden =
        hentUtKvartal(datagrunnlag, ettÅrSiden);

    if (nyesteSykefravær.isEmpty() || sykefraværetEttÅrSiden.isEmpty()) {
      return Either.left(
          new UtilstrekkeligDataException(
              "Mangler data for " + sistePubliserteKvartal + " og/eller " + ettÅrSiden));
    }

    Either<StatistikkException, BigDecimal> nyesteSykefraværsprosent
        = nyesteSykefravær.get().kalkulerSykefraværsprosent();
    Either<StatistikkException, BigDecimal> sykefraværsprosentEttÅrSiden
        = sykefraværetEttÅrSiden.get().kalkulerSykefraværsprosent();

    if (nyesteSykefraværsprosent.isLeft() || sykefraværsprosentEttÅrSiden.isLeft()) {
      return Either.left(new UtilstrekkeligDataException(
          "Feil i utregningen av sykefraværsprosenten, kan ikke regne ut trendverdi."));
    }

    BigDecimal trendverdi = nyesteSykefraværsprosent.get()
        .subtract(sykefraværsprosentEttÅrSiden.get());

    int antallTilfeller =
        nyesteSykefravær.get().getAntallPersoner()
            + sykefraværetEttÅrSiden.get().getAntallPersoner();

    return Either.right(
        new Trend(trendverdi, antallTilfeller, List.of(sistePubliserteKvartal, ettÅrSiden)));
  }
}

File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/applikasjon/dataTilFrontends/Trendkalkulator.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.dataTilFrontends

import io.vavr.control.Either
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domene.Trend
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domene.ÅrstallOgKvartal
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.UmaskertSykefraværForEttKvartal
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.UtilstrekkeligDataException

data class Trendkalkulator(
    var datagrunnlag: List<UmaskertSykefraværForEttKvartal>? = null,
    var sistePubliserteKvartal: ÅrstallOgKvartal? = null
) {
    fun kalkulerTrend(): Either<UtilstrekkeligDataException, Trend> {
        val ettÅrSiden = sistePubliserteKvartal!!.minusEttÅr()
        val nyesteSykefravær = UmaskertSykefraværForEttKvartal.hentUtKvartal(datagrunnlag, sistePubliserteKvartal!!)
        val sykefraværetEttÅrSiden = UmaskertSykefraværForEttKvartal.hentUtKvartal(datagrunnlag, ettÅrSiden)
        if (nyesteSykefravær.isEmpty || sykefraværetEttÅrSiden.isEmpty) {
            return Either.left(
                UtilstrekkeligDataException(
                    "Mangler data for $sistePubliserteKvartal og/eller $ettÅrSiden"
                )
            )
        }
        val nyesteSykefraværsprosent = nyesteSykefravær.get().kalkulerSykefraværsprosent()
        val sykefraværsprosentEttÅrSiden = sykefraværetEttÅrSiden.get().kalkulerSykefraværsprosent()
        if (nyesteSykefraværsprosent.isLeft || sykefraværsprosentEttÅrSiden.isLeft) {
            return Either.left(
                UtilstrekkeligDataException(
                    "Feil i utregningen av sykefraværsprosenten, kan ikke regne ut trendverdi."
                )
            )
        }
        val trendverdi = nyesteSykefraværsprosent.get().subtract(sykefraværsprosentEttÅrSiden.get())
        val antallTilfeller = (nyesteSykefravær.get().antallPersoner
                + sykefraværetEttÅrSiden.get().antallPersoner)
        return Either.right(
            Trend(trendverdi, antallTilfeller, java.util.List.of(sistePubliserteKvartal, ettÅrSiden))
        )
    }
}